packages/cursorless-engine/src/core/commandVersionUpgrades/generateSpokenForm.ts

```diff
@@ -0,0 +1,478 @@
+  return [length, scopePlural, connectives.backward];
+}
+
+function relativeScopeExclusiveToSpokenForm(
+  modifier: RelativeScopeModifier,
+): RecursiveArray<string> {
+  const scope = scopeTypeToSpokenForm(modifier.scopeType);
+  const direction =
+    modifier.direction === "forward" ? connectives.next : connectives.previous;
+
+  if (modifier.offset === 1) {
+    const number = numberToSpokenForm(modifier.length);
+
+    if (modifier.length === 1) {
+      // next/previous token
+      return [direction, scope];
+    }
+
+    const scopePlural = pluralize(scope);
+
+    // next/previous two tokens
+    return [direction, number, scopePlural];
+  }
+
+  if (modifier.length === 1) {
+    const ordinal = ordinalToSpokenForm(modifier.offset);
+    // second next/previous token
+    return [ordinal, direction, scope];
+  }
+
+  throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+}
+
+function scopeTypeToSpokenForm(scopeType: ScopeType): string {
+  switch (scopeType.type) {
+    case "oneOf":
+    case "customRegex":
+    case "switchStatementSubject":
+      throw new NoSpokenFormError(`Scope type '${scopeType.type}'`);
+    case "string":
+      throw new AmbiguousSpokenFormError(`Scope type '${scopeType.type}'`);
+    case "surroundingPair": {
+      const pair = surroundingPairNameToSpokenForm(scopeType.delimiter);
+      if (scopeType.forceDirection != null) {
+        const direction =
+          scopeType.forceDirection === "left"
+            ? surroundingPairForceDirections.left
+            : surroundingPairForceDirections.right;
+        return `${direction} ${pair}`;
+      }
+      return pair;
+    }
+
+    default:
+      return scopeSpokenForms[scopeType.type];
+  }
+}
+
+function markToSpokenForm(mark: PartialMark): RecursiveArray<string> {
+  switch (mark.type) {
+    case "decoratedSymbol": {
+      const [color, shape] = mark.symbolColor.split("-");
+      const components: string[] = [];
+      if (color !== "default") {
+        components.push(hatColorToSpokenForm(color));
+      }
+      if (shape != null) {
+        components.push(hatShapeToSpokenForm(shape));
+      }
+      components.push(characterToSpokenForm(mark.character));
+      return components;
+    }
+
+    case "lineNumber": {
+      return lineNumberToParts(mark);
+    }
+
+    case "range": {
+      if (
+        mark.anchor.type === "lineNumber" &&
+        mark.active.type === "lineNumber"
+      ) {
+        const [typeAnchor, numberAnchor] = lineNumberToParts(mark.anchor);
+        const [typeActive, numberActive] = lineNumberToParts(mark.active);
+        if (typeAnchor === typeActive) {
+          const connective = getRangeConnective(
+            mark.excludeAnchor ?? false,
+            mark.excludeActive ?? false,
+          );
+          // Row five past seven
+          return [typeAnchor, numberAnchor, connective, numberActive];
+        }
+      }
+
+      throw Error(`Mark '${mark.type}' is not fully implemented`);
```

Comment:
Once again, I think I'd throw `NoSpokenFormError`, no?

Reply:
This one I'm slightly more on the fence about, but again I think mostly the problem is just that we don't currently have a spoken form for it. I don't think there's anything inherently wrong with this case, and we do support it extension-side. But again this one's a bit different because spoken form would prob be ambiguous and we may remove this construct in the future. But I'm still not sure I see the benefit of not just throwing `NoSpokenFormError`. I think I'd reserve `Error` for when there's truly something wrong / unexpected. But if you want to push the other direction for this specific case I could be convinced